### PR TITLE
BEAM-13592 Add getOrderingKey in o.a.b.sdk.io.gcp.pubsub.PubsubMessage

### PR DESCRIPTION
**Pull request by gemelen**
Add orderingKey from proto PubsubMessage to beam Pubsubmessage.
Add a new coder for a PubsubMessage with all fields.
Propagate this new view of a PubsubMessage to PubsubIO as a new API
call.
Add a unit test for a coder.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by gemelen** [comment]
Solves (hopefully) 
- https://github.com/apache/beam/issues/21304

**Comment by gemelen** [comment]
@damccorm Hi, could you please suggest a reviewer for this changeset?

**Comment by gemelen** [comment]
Addresses BEAM-13592

**Comment by damccorm** [comment]
@johnjcasey who is a good reviewer here?

**Comment by johnjcasey** [comment]
I am, I'll take a look

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java:533]: Can we remove "all" from the message name to keep it consistent with the above method?
Also, based on the above method, do we need .setNeedsAttributes(true) here?
Finally, should we name the coder PubsubMessageCoder, given that the above method calls it PubsubMessageWithAttributesAndMessageIdCoder?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:42]: can the attribute map be null, or can elements within the map be null?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:64]: What needs to be discussed here?

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java
Comment (gemelen) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:42]: This part is identical to other coders, so I assume that attributes field either null or contains something.
My guess that elements could be null too, but this needs to be changed in all PubsubMessage-related coders.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java
Comment (gemelen) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java:533]: I think that since this coder grabs all fields of an incoming PubsubMessage then it could be named without specifying all these fields in its name. Accordingly, method would be better named 'readMessages` but this name is already taken by other implementation (that reads only payload of PubsubMessage).

I agree that `readMessagesWithAllAttributesAndMessageIdAndOrderingKey` then should be named `readMessagesWithAttributesAndMessageIdAndOrderingKey`, so I'm updating it to this.

Regardint the
> Also, based on the above method, do we need .setNeedsAttributes(true) here?

newly added method doesn't include `setNeedsAttributes`, it's a `setNeedsOrderingKey`:

```
public static Read<PubsubMessage> readMessagesWithAllAttributesAndMessageIdAndOrderingKey() 
{
    return Read.newBuilder()
     .setCoder(PubsubMessageCoder.of())
     .setNeedsOrderingKey(true)
     .build();
}
```

**Comment by gemelen** [comment]
@johnjcasey could you please take a look on a changeset once more

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessages.java
Comment (gemelen) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessages.java:46]: It looks like it does, I guess.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java
Comment (egalpin) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java:527]: I believe to support bounded writes, on line 1329 of this file you'll need to add explicit copying of the orderingKey as well (in the `output.add(OutgoingMessage.of(...` section of `processElement`

**Comment by egalpin** [comment]
Relates to #21162.  I think it may also be worth mentioning that there was a previous email thread[1] about the feature to support subscriptions where ordering is enabled, but Beam itself does not guarantee ordering so this could be a confusing user experience.  I may be missing additional context which has developed since that 2-year-old thread though!

Writes with orderingKey should definitely be supported though :-) 

[1] https://lists.apache.org/thread/1koqhmhr1tb5572cj3fq5b65dnwbg7fg

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java
Comment (egalpin) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:38]: I noticed that the other coders, like `PubsubMessagePayloadOnlyCoder` have specific names indicating the properties of the messages which they are meant to be the Coder for.  IMO the name `PubsubMessageCoder` implies by name that it's a generic coder for pubsub messages.  Following the naming convention in place, this class might be named `PubsubMessageWithAttributesAndMessageIdAndOrderingKeyCoder` (which is comically long, so I'm not sure which is better haha).

Comment (egalpin) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:38]: I believe you'll also need to register this in [PubsubCoderProviderRegistrar](https://github.com/apache/beam/blob/bc97c6b90863f18e42070287e418df2b2d14f041/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java)

Comment (egalpin) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageCoder.java:38]: Which I think actually implies that there will need to be a coder for all possible permutations (payload, messageId, attributes, orderingKey)

**Comment by egalpin** [comment]
At the risk of hijacking... Having support for orderingKey is high-priority for me as well, so I've been playing around with this locally.  I can't for the life of me get it to work with dataflow when attempting to publish messages with an ordering key.  I did get it to work locally, where I could see (via the GCP UI) published messages had orderingKey assigned.  Something must be overriding it in dataflow.  Thought it might be worth mentioning.

I also had a lot of trouble with coders and getting the correct coder to be maintained throughout the publish process, specifically the coder would be altered to be `PubsubMessageWithAttributesCoder` after data went through [PubsubMessages.ParsePayloadAsPubsubMessageProto](https://github.com/apache/beam/blob/367173f6245f8df514b2e0526ab659665344f5a6/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java#L414-L415).  I'm definitely missing an understanding of how to make proper use of CoderRegistrar

**Comment by egalpin** [comment]
here's a comparison showing some of the changes I was playing with:  https://github.com/gemelen/beam/compare/beam-13592-pubsub-java-orderingkey...egalpin:beam:egalpin/gemelen-beam-13592-fork?expand=1

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java
Comment (egalpin) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java:34]: I'm not 100% confident in the order that is required here, or if order matters.  I couldn't find documentation around how coders are resolved when multiple coders are registered for the same class.  I tried tracing the code but wasn't able to come up with a definitive answer

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java
Comment (egalpin) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java:41]: moved this code to be the last in the registrar (rather than first), otherwise there are failures in the tests.  @johnjcasey Do you have any further info on coder registrars and how a coder is resolved when multiple coders are registered for the same class?

**Comment by gemelen** [comment]
@damccorm Could you please help to find a reviewer?

**Comment by damccorm** [comment]
@johnjcasey mind taking another look?

**Review by johnjcasey — APPROVED** [review]
LGTM

**Comment by johnjcasey** [comment]
Though it would be good to get a better understanding of coders first. I think @TheNeuralBit has some more experience there

**Comment by TheNeuralBit** [comment]
> Though it would be good to get a better understanding of coders first. I think @TheNeuralBit has some more experience there

Is there a specific question I can help with?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java
Comment (egalpin) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java:41]: @TheNeuralBit I think this is the area where there is lack of clarity.  As best as I can tell, here there are multiple coders registered for the same class (`PubsubMessage.class`).  It seems like order matters here, but I'm not sure all the implications.  Ex. if I put the new coder first in the registrar, existing pubsub unit tests will fail with coder errors.

Comment (egalpin) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java:41]: Thanks for the info @TheNeuralBit , this is super helpful context.  Sounds like what's here then is not going to break anything, which is the most important.

One aspect that I predict will be confusing for users is that if one attempts to _write_ pubsub messages with ordering key, the `PubsubMessageWithAttributesCoder` will be used by default unless `setCoder` is called; in that default case, the orderingKey will be "mysteriously" stripped away by the coder.  I suppose we might also need a new `write` helper method like `writeMessagesWithOrderingKeys` in addition to `writeMessages` in order help users to set the correct coder?

**Comment by egalpin** [comment]
Friendly bump, anything outstanding here or is this ready to be merged? (cc @johnjcasey @TheNeuralBit )

**Comment by TheNeuralBit** [comment]
> Friendly bump, anything outstanding here or is this ready to be merged? (cc @johnjcasey @TheNeuralBit )

Apologies. I think this is ready to merge once the merge conflict is resolved.

I will file an issue to track the default coder issue

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java
Comment (TheNeuralBit) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubCoderProviderRegistrar.java]: Filed https://github.com/apache/beam/issues/23525 for the follow-up

**Comment by egalpin** [comment]
thanks @TheNeuralBit !  I merged in master + resolved conflicts.  For future reference, is merge or rebase preferred here? I opted for merge so that prior commits would not be marked as new for reviewers.

**Comment by egalpin** [comment]
Thank you @gemelen for kicking off this high-value addition to PubsubIO! 🎉

**Comment by gemelen** [comment]
> Thank you @gemelen for kicking off this high-value addition to PubsubIO! 🎉

You deserve thank you for that changeset more than anybody else.

**Comment by gemelen** [comment]
Oh, last thing is to update https://issues.apache.org/jira/browse/BEAM-13592

**Comment by TheNeuralBit** [comment]
> thanks @TheNeuralBit ! I merged in master + resolved conflicts. For future reference, is merge or rebase preferred here? I opted for merge so that prior commits would not be marked as new for reviewers.

I think this is just up to personal preference :)

**Pull request merged**
